designs/disruption-controls.md

```diff
@@ -0,0 +1,118 @@
+
+## Proposed Spec
+
+This only includes the `Budgets` and `ConsolidateAfter` portion of the disruption spec.
+
+```{yaml}
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  consolidationPolicy: WhenUnderutilized || WhenEmpty
+  consolidateAfter: 10m || Never # metav1.Duration
+  disruption:
+    budgets:
+    # On Weekdays during business hours, don't do any deprovisioning.
+    - crontab: "0 9 * * mon-fri"
+      duration: 8h
+      maxUnavailable: 0
+    # Every other time, only allow 10 nodes to be deprovisioned simultaneously
+    - maxUnavailable: 10
+```
+
+## Code Definition
+
+```{go}
+type Disruption struct {
+    {...}
+    // Budgets is a list of Budgets.
+    // If there are multiple active budgets, the most restrictive is respected.
+    Budgets []Budget `json:"budgets,omitempty" hash:"ignore"`
+    // ConsolidateAfter is a nillable duration, parsed as a metav1.Duration.
+    // Users can use "Never" to disable Consolidation.
+    ConsolidateAfter *NillableDuration `json:"consolidateAfter" hash:"ignore"`
+    // ConsolidationPolicy determines how Karpenter will consider nodes
+    // as candidates for Consolidation.
+    // WhenEmpty uses the same behavior as v1alpha5 TTLSecondsAfterEmpty
+    // WhenUnderutilized uses the same behavior as v1alpha5.ConsolidationEnabled: true
+    ConsolidationPolicy string `json:"consolidationPolicy" hash:"ignore"`
+}
+// Budget specifies periods of times where Karpenter will restrict the
+// number of Node Claims that can be terminated at a time.
+// Unless specified, a budget is always active.
+type Budget struct {
+    // MaxUnavailable dictates how many Node Claims owned by this NodePool
+    // can be terminating at once.
+    // This only respects and considers nodes with a DeletionTimestamp set.
+    MaxUnavailable intstr.IntOrString `json:"maxUnavailable" hash:"ignore"`
+    // Crontab specifies when a budget begins being active.
+    // Crontab uses the same syntax as a Cronjob:
+    // "Minute Hour DayOfMonth Month DayOfWeek"
+    // This is required if Duration is set.
+    Crontab *string `json:"crontab,omitempty" hash:"ignore"`
+    // Duration determines how long a Budget is active since each Crontab hit.
+    // This is required if Crontab is set.
+    Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+}
+```
+
+## Validation/Defaults
+
+For each `Budget`, users must set a non-negative `MaxUnavailable`. Users can disable scale down for a NodePool by setting this to `0`. Users must either omit both `Crontab` and `Duration` or set both of them, since `Crontab` and `Duration` are inherently linked. Omitting these two fields will be equivalent to an always active `Budget`. If a user defines seconds for `Duration`, Karpenter will ignore it, effectively rounding down to the nearest minute, since the smallest denomination of time in upstream Crontabs are minutes.
```

Comment:
If we're not going to respect seconds, we should just not allow them to be specified within our regex